### PR TITLE
chore: Add the support of fetching game info from backend on forked ExoPlayer

### PR DESCRIPTION
**Pull request by guoen21**
https://dicetech.atlassian.net/browse/DORIS-1716

We can config the test stream as following on assets/media.exolist.json of demo app:
```
      {
        "name": "CEA708 live, wwe",
        "uri": "https://dummy.imggaming.com/video?id=111449&live=true&dash=&realm=dce.wwe&usr=&pwd="
      },
      {
        "name": "450830, DRM, adjara",
        "uri": "https://dummy.imggaming.com/video?id=450830&live=&dash=true&realm=dce.adjara&usr=&pwd="
      },
      ...
```
The stream url should contain `dummy.imggaming.com` as host name, it is the key word.
id -  the dice id of this test stream
live - this test stream is live or vod
dash - play the hls stream or dash stream
usr, pwd - the name and password for sign in

### CODE REVIEW

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/EntityUtil.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import com.google.android.exoplayer2.demo.upstream.entity.StreamResponse;
+
+public class EntityUtil {
+
+  private EntityUtil() {
+  }
+
+  // Util methods.
+
+  public static ProtocolData getProtocolData(boolean forceDash, StreamResponse result) {
```
Comment (szaboa) [inline demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/EntityUtil.java]: Maybe add these 3 static methods to `ProtocolData` itself? then we can get rid of this class

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/ClientUtil.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package com.google.android.exoplayer2.demo.upstream;
+
+import java.util.concurrent.TimeUnit;
+import okhttp3.OkHttpClient;
+
+public class ClientUtil {
```
Comment (szaboa) [inline demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/ClientUtil.java]: maybe rename to `OkHttpClientUtil` or `OkHttpClientFactory` ?

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/PlaybackAsyncAPI.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  })
+  @GET
+  Observable<PlaybackResponse> getPlaybackUrl(
+      @Url String url,
+      @Header("realm") String realm,
+      @Header("authorization") String authorizationToken
+  );
+
+  @GET
+  Observable<StreamResponse> getStreamInfo(
```
Comment (szaboa) [inline demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/PlaybackAsyncAPI.java:45]: maybe rename to `getLiveStreamInfo`?

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/SampleChooserActivity.java [hunk 4]
```diff
@@ -239,6 +242,31 @@ public boolean onChildClick(
     intent.putExtra(
         IntentUtil.PREFER_EXTENSION_DECODERS_EXTRA,
         isNonNullAndChecked(preferExtensionDecodersMenuItem));
+    if (playlistHolder.mediaItems.size() == 1) {
+      AtomicInteger appendCount = new AtomicInteger(0);
+      Uri originUri = playlistHolder.mediaItems.get(0).localConfiguration.uri;
+      PlaybackProvider.getInstance().getStreamInfo(originUri)
+          .observeOn(AndroidSchedulers.mainThread())
+          .subscribe(result -> {
```
Comment (szaboa) [inline demos/main/src/main/java/com/google/android/exoplayer2/demo/SampleChooserActivity.java:248]: I think we could simply the logic by removing the `onComplete`. 
If we return an `Observable.error(..)` in `PlaybackProvider#getStreamInfo` if the media is null, then we will have only two cases:

**onNext:**
```
IntentUtil.addToIntent(Collections.singletonList(result), intent);
startActivity(intent);
```

**onError:**
```
IntentUtil.addToIntent(playlistHolder.mediaItems, intent);
startActivity(intent);
```

**Review by szaboa — DISMISSED** [review]

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/EntityUtil.java (hunk 1 above)
Comment (guoen21) [reply demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/EntityUtil.java]: done

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/ClientUtil.java (hunk 2 above)
Comment (guoen21) [reply demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/ClientUtil.java]: done

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/SampleChooserActivity.java (hunk 4 above)
Comment (guoen21) [reply demos/main/src/main/java/com/google/android/exoplayer2/demo/SampleChooserActivity.java:248]: For the case of media is null, now it return empty `MediaItem` instance (It is not an error) and then we also can remove the `onComplete()`.

File: demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/PlaybackAsyncAPI.java (hunk 3 above)
Comment (guoen21) [reply demos/main/src/main/java/com/google/android/exoplayer2/demo/upstream/PlaybackAsyncAPI.java:45]: done

**Review by szaboa — APPROVED** [review]

**Pull request merged**